### PR TITLE
LTI 1.3 assessment

### PR DESCRIPTION
**Pull request by trombonekenny**
- Adds `lineitems` support to LTI 1.3 course instances
- Instructor course instance LTI 1.3 admin pages UX to include assessments

### CODE REVIEW

**Comment by trombonekenny** [comment]
Pulling out of draft to review, but there's a few more things I want to check on before merging.

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 1]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+  });
+
+  job.info('Associating PrairieLearn assessment with the new assignment');
+
+  await link_assessment(instance.lti13_course_instance.id, assessment.id, item);
+  job.info('Done.');
+}
+
+export async function query_and_link_lineitem(
+  instance: any,
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.ts:401]: Better types please!

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 2]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+    lti13_course_instance_id,
+    lineitem_id: lineitem.id,
+    lineitem: JSON.stringify(lineitem),
+    assessment_id,
+  });
+}
+
+function sleep(delay: number) {
+  return new Promise((resolve) => setTimeout(resolve, delay));
+}
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.ts:440]: https://github.com/PrairieLearn/PrairieLearn/blob/4f1e6a8f325b33ea23fed227d0d035ccdea0e116/apps/prairielearn/src/cron/index.ts#L1

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 3]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+
+/* Throttling notes
+// https://canvas.instructure.com/doc/api/file.throttling.html
+// 403 Forbidden (Rate Limit Exceeded)
+// X-Request-Cost
+// X-Rate-Limit-Remaining
+*/
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.ts:447]: What are these notes relevant to? I don't see any special handling of either the status code or headers.

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 4]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+
+export async function fetchRetry(
+  input: RequestInfo | URL,
+  opts?: RequestInit | undefined,
+  incomingfetchRetryOpts?,
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.ts:452]: Types!

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 5]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+      }
+    }
+
+    return await response.json();
+  } catch (err) {
+    fetchRetryOpts.retryLeft -= 1;
+    if (fetchRetryOpts.retryLeft === 0) {
+      throw err;
+    }
+    await sleep(fetchRetryOpts.sleepMs);
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.ts:481]: Did you consider using exponential backoffs?

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql [hunk 6]
```diff
@@ -11,10 +11,138 @@ WHERE
...
+SELECT
   *
+FROM
+  lti13_lineitems
+WHERE
+  lti13_course_instance_id = $lti13_course_instance_id
+ORDER BY
+  lineitem_id;
+
+-- BLOCK select_assessment_to_create
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql:102]: Is "link" a better verb here than "create"?

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts [hunk 7]
```diff
@@ -87,3 +333,50 @@ export function InstructorInstanceAdminLti13({
...
+      class="btn btn-xs"
+      onClick="event.preventDefault();$(this).next('.lineitem-detail').toggle();"
+    >
+      ...
+    </button>
+    <div class="lineitem-detail" style="display:none;">
+      <pre>${JSON.stringify(item, null, 2)}</pre>
+    </div>
+  `;
+  */
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts:352]: 👀

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts [hunk 8]
```diff
@@ -87,3 +333,50 @@ export function InstructorInstanceAdminLti13({
...
+}
+
+export function LineitemsInputs(lineitems: Lti13Lineitem[]) {
+  if (lineitems.length === 0) {
+    return html`<p>None found.</p>`.toString();
+  }
+  // TODO: Filter out assignments that are already linked? Re-linking will override the
+  // previous assignment link
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts:360]: Do you plan to address this before this PR lands?

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts [hunk 9]
```diff
@@ -61,11 +88,230 @@ export function InstructorInstanceAdminLti13({
...
+                                <a href="${urlPrefix}/assessment/${row.id}/"
+                                  >${row.title}
+                                  ${row.group_work
+                                    ? html` <i class="fas fa-users" aria-hidden="true"></i> `
+                                    : ''}</a
+                                >
+                              </td>
+                              <td>
+                                ${Modal({
+                                  body: html`<p>Which ${lms_name} assignment should we link?</p>
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts:214]: Nit:

```suggestion
                                  body: html`
                                    <p>Which ${lms_name} assignment should we link?</p>
```

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.ts [hunk 10]
```diff
@@ -58,6 +100,23 @@ router.get(
...
+      },
+      Lti13CombinedInstanceSchema,
+    );
+
+    const serverJobOptions = {
+      courseInstanceId: res.locals.course_instance.id,
+      userId: res.locals.user.user_id,
+      authnUserId: res.locals.authn_user.user_id,
+      type: 'lti13',
+      description: 'Some LTI operation',
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.ts:117]: Consider making this more exact?

**Review by nwalters512 — COMMENTED** [review]
First pass. I haven't yet fired this up with a local Canvas install to test any behavior/UI/UX/etc.

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 5 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:481]: I did, but the note at https://canvas.instructure.com/doc/api/file.throttling.html made me feel like a more complicated backoff would just be overly complex.

> Since the cost of a request is roughly based on the amount of time it takes to process, and the quota (by default) replenishes at a rate faster than real-time, any API client that makes no more than one simultaneous request is unlikely to be throttled.

We know we're likely to queue these kinds of commands anyway, so I didn't make this too complicated for a MVP.

File: apps/prairielearn/src/ee/lib/lti13.sql [hunk 11]
```diff
@@ -8,3 +8,99 @@ SELECT
...
+  assessment_id = $assessment_id;
+
+-- BLOCK create_lineitems_temp
+CREATE TEMPORARY TABLE new_lineitems (
+  lti13_course_instance_id bigint,
+  lineitem_id text,
+  lineitem jsonb
+) ON
+COMMIT
+DROP;
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.sql:48]: It's part of the sync logic that follows.

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 12]
```diff
@@ -1,14 +1,55 @@
...
 
 const sql = loadSqlEquiv(import.meta.url);
 
+const TOKEN_SCOPES = [
+  'https://purl.imsglobal.org/spec/lti-ags/scope/lineitem',
+  //'https://purl.imsglobal.org/spec/lti-ags/scope/result.readonly',
+  'https://purl.imsglobal.org/spec/lti-ags/scope/score',
+  //'https://purl.imsglobal.org/spec/lti-nrps/scope/contextmembership.readonly',
+  //'https://purl.imsglobal.org/spec/lti-ags/scope/lineitem.readonly',
+  //'https://canvas.instructure.com/lti/public_jwk/scope/update',
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:28]: The non-commented lines are the scope we currently need to fulfill the functionality so far. We'll likely have to do something with contextmembership.readonly in the next part, so I left the full list in to make it easier to add them later (if we discover we need them). Or I can remove them.

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 13]
```diff
@@ -1,14 +1,55 @@
...
+];
+
+export const Lti13CombinedInstanceSchema = z.object({
+  lti13_course_instance: Lti13CourseInstanceSchema,
+  lti13_instance: Lti13InstanceSchema,
+});
+export type Lti13CombinedInstance = z.infer<typeof Lti13CombinedInstanceSchema>;
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:35]: I actually use this in the lti13.ts library in a few places where you caught that I was still using `any`.

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 14]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+  instance: any,
+  job: ServerJob,
+  assessment: {
+    label: string;
+    id: string;
+    url: string;
+  },
+) {
+  const createBody: Lti13Lineitem = {
+    id: 'new_lineitem', // will be ignored/overwritten by the LMS platform
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:367]: It was to make the typing work. TS didn't like me making a `Lti13Lineitem` object without an ID to send to the POST API.
https://canvas.instructure.com/doc/api/line_items.html#method.lti/ims/line_items.create

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 15]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+    scoreMaximum: 100,
+    label: assessment.label,
+    resourceId: assessment.id,
+    // tag ???
+    // startDateTime null
+    // endDateTime null
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:373]: Not immediately, no, but it's a discussion started of if we want to use something like this.

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 3 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:447]: `response.ok` evaluates to true if 200 <= response status < 300. The current version of the code doesn't trigger off of 403 differently than any other 400 or 500 error (it retries), but those notes were the beginning of that logic.

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.ts (hunk 10 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.ts:117]: This is the boilerplate that gets overwritten with more exact descriptions/details based on the __action.

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql [hunk 16]
```diff
@@ -0,0 +1,11 @@
+ALTER TABLE lti13_course_instances
+ADD COLUMN IF NOT EXISTS lineitems TEXT;
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql:2]: I tried be consistent with friendly naming in the UI and spec naming in the database or functions. Yes, it's a URL: https://www.imsglobal.org/spec/lti-ags/v2p0#services-definition

I did consider naming it `ags_lineitems` but `lineitems` should be specific enough in the LTI 1.3 space that it wouldn't overlap?

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 17]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+}
+
+/*
+ * Canvas supports an API call to get a single lineitem.
+ * https://canvas.instructure.com/doc/api/line_items.html#method.lti/ims/line_items.show
+ * But in testing Dave discovered its output might not be the same as
+ * lineitems (plural).
+ *
+ * Instead of a get_lineitem() function, let's get_lineitems and filter.
+ */
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:316]: On further investigation, I decided to remove this and go ahead and make a getLineitem() function.

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:28]: Changed in 6cb43ef48

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts:35]: In 6cb43ef48 I left it here, but referenced it in `lib/lti13.ts`

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 18]
```diff
@@ -211,3 +257,223 @@ export async function validateLti13CourseInstance(
     z.boolean(),
   );
 }
+
+export async function access_token(lti13_instance_id: string) {
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts]: Resolved in 6cb43ef48

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 19]
```diff
@@ -211,3 +257,223 @@ export async function validateLti13CourseInstance(
...
+  const lti13_instance = await selectLti13Instance(lti13_instance_id);
+
+  let tokenSet: TokenSet = lti13_instance.access_tokenset;
+
+  if (
+    lti13_instance.access_token_expires_at &&
+    lti13_instance.access_token_expires_at > new Date()
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts]: Added a 5 minute buffer in 6cb43ef48

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 20]
```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
...
+    return tokenSet.access_token;
+  }
+}
+
+export async function get_lineitems(instance: Lti13CombinedInstance) {
+  if (instance.lti13_course_instance.lineitems == null) {
+    throw new Error('Lineitems not defined');
+  }
+  const token = await access_token(instance.lti13_instance.id);
+  const lineitems: Lti13Lineitem[] = await fetchRetry(instance.lti13_course_instance.lineitems, {
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts]: Added `parse()` in 6cb43ef48

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts]: Resolved in 6cb43ef48

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql (hunk 6 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql:102]: create_and_link? I'm inclined to leave it as-is.

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts (hunk 7 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts]: Removed in bf564fc79

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts (hunk 8 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts:360]: Probably not. Enhancing the displayed list of learning outcomes to link can come later.

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql [hunk 21]
```diff
@@ -0,0 +1,11 @@
...
+
+CREATE TABLE IF NOT EXISTS lti13_lineitems (
+  lti13_course_instance_id BIGINT REFERENCES lti13_course_instances ON DELETE CASCADE ON UPDATE CASCADE,
+  assessment_id BIGINT NOT NULL REFERENCES assessments ON DELETE CASCADE ON UPDATE CASCADE,
+  lineitem_id text,
+  lineitem jsonb,
+  last_activity timestamp with time zone,
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql]: Add Last activity display in bf564fc79

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql [hunk 22]
```diff
@@ -0,0 +1,11 @@
+ALTER TABLE lti13_course_instances
+ADD COLUMN IF NOT EXISTS lineitems TEXT;
+
+CREATE TABLE IF NOT EXISTS lti13_lineitems (
+  lti13_course_instance_id BIGINT REFERENCES lti13_course_instances ON DELETE CASCADE ON UPDATE CASCADE,
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql]: Updated the database structure in bf564fc79

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+  assessment_id BIGINT NOT NULL REFERENCES assessments ON DELETE CASCADE ON UPDATE CASCADE,
+  lineitem_id text,
+  lineitem jsonb,
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql]: Updated the database structure in bf564fc79

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql [hunk 24]
```diff
@@ -0,0 +1,11 @@
...
+  last_activity timestamp with time zone,
+  PRIMARY KEY (lti13_course_instance_id, lineitem_id)
```
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql]: Updated the database structure in bf564fc79

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/ee/lib/lti13.ts]: Hmm, why not include all the scopes we reasonably think we'd need in the near future? That seems preferable to having to go back to every LTI admin to request additional scopes in the future (I imagine that's what we'd have to do otherwise, right?).

File: apps/prairielearn/src/ee/lib/lti13.sql (hunk 11 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/ee/lib/lti13.sql:51]: I understand that it's used in the syncing process. I'll rephrase the question: why does the syncing logic _require_ a temporary table? We basically never do that anywhere else in PL.

File: apps/prairielearn/src/ee/lib/lti13.ts [hunk 25]
```diff
@@ -211,3 +261,232 @@ export async function validateLti13CourseInstance(
...
+
+  let tokenSet: TokenSet = lti13_instance.access_tokenset;
+
+  // -5 minute buffer to refresh tokens before they expire
+  if (
+    lti13_instance.access_token_expires_at &&
+    lti13_instance.access_token_expires_at > new Date(Date.now() - 5 * 60 * 1000)
+  ) {
+    return tokenSet.access_token;
+  } else {
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.ts:276]: Nit: with the early return above, you can drop the `else` to reduce indentation.

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts [hunk 26]
```diff
@@ -87,3 +339,41 @@ export function InstructorInstanceAdminLti13({
     </html>
   `.toString();
 }
+
+function lineItem(item: Lti13Lineitems, timezone: string) {
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.html.ts:343]: Conventionally, this should be named `LineItem`.

File: apps/prairielearn/src/lib/db-types.ts [hunk 27]
```diff
@@ -746,6 +747,16 @@ export const LtiCredentialsSchema = z.object({
 });
 export type LtiCredentials = z.infer<typeof LtiCredentialsSchema>;
 
+export const Lti13LineitemsSchema = z.object({
+  id: IdSchema,
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/db-types.ts:751]: Let's keep column names alphabetized.

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql [hunk 28]
```diff
@@ -0,0 +1,12 @@
+ALTER TABLE lti13_course_instances
+ADD COLUMN IF NOT EXISTS lineitems TEXT;
+
+CREATE TABLE IF NOT EXISTS lti13_lineitems (
+  id BIGSERIAL PRIMARY KEY,
+  assessment_id BIGINT NOT NULL UNIQUE REFERENCES assessments ON DELETE CASCADE ON UPDATE CASCADE,
```
Comment (nwalters512) [inline apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql:6]: Re-raising for discussion from another resolved comment: should we add a constraint forbidding multiple line items for the same assessment? We can discuss at the dev meeting this week if that's best.

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/ee/lib/lti13.ts:376]: I'd recommend changing the types to match reality, rather than code to match the types. You can type this as `Omit<Lti13LineItem, 'id'>` or something like that.

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 5 above)
Comment (nwalters512) [reply apps/prairielearn/src/ee/lib/lti13.ts:489]: Famous last words 😉

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql (hunk 28 above)
Comment (trombonekenny) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql:6]: Doesn't the UNIQUE there do that?

File: apps/prairielearn/src/lib/db-types.ts (hunk 27 above)
Comment (trombonekenny) [reply apps/prairielearn/src/lib/db-types.ts]: Fixed in 46c87e028

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 25 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts]: Addressed in 46c87e028

File: apps/prairielearn/src/ee/lib/lti13.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.ts]: Addressed in 46c87e028

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql (hunk 28 above)
Comment (nwalters512) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql:6]: Doh, I'm not used to looking out for `UNIQUE` inline like this. You're absolutely right.

File: apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql (hunk 16 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/migrations/20240712152823_lti13_course_instances__add_lineitems.sql]: This was changed to `lineitems_url` in a recent commit.

File: apps/prairielearn/src/ee/lib/lti13.sql [hunk 29]
```diff
@@ -8,3 +8,89 @@ SELECT
...
+
+-- BLOCK update_token
+UPDATE lti13_instances
+SET
+  access_tokenset = $access_tokenset,
+  access_token_expires_at = $access_token_expires_at
+WHERE
+  id = $lti13_instance_id;
+
+-- BLOCK update_lineitem
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.sql:20]: Nit: this query name reflects the old table name. This should be named something like `upsert_lti13_assessment`. This can be addressed in a future PR.

File: apps/prairielearn/src/ee/lib/lti13.sql [hunk 30]
```diff
@@ -8,3 +8,89 @@ SELECT
...
+    NOW()
+  )
+ON CONFLICT (lti13_course_instance_id, lineitem_id_url) DO
+UPDATE
+SET
+  lineitem = $lineitem,
+  assessment_id = $assessment_id,
+  last_activity = NOW();
+
+-- BLOCK sync_lti13_lineitems
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.sql:44]: This query name reflects the old table name.

File: apps/prairielearn/src/ee/lib/lti13.sql [hunk 31]
```diff
@@ -8,3 +8,89 @@ SELECT
...
+      updating
+  ) as updated,
+  (
+    SELECT
+      count(*)
+    FROM
+      deleting
+  ) AS deleted;
+
+-- BLOCK delete_lineitem_by_assessment_id
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/lti13.sql:92]: Same comment.

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql [hunk 32]
```diff
@@ -11,10 +11,138 @@ WHERE
...
+  (
+    CASE
+      WHEN $assessments_group_by = 'Module' THEN am.id
+    END
+  ),
+  aset.number,
+  a.order_by,
+  a.id;
+
+-- BLOCK select_lineitems
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql:92]: Old table name.

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql [hunk 33]
```diff
@@ -11,10 +11,138 @@ WHERE
...
+WHERE
+  a.course_instance_id = $course_instance_id
+  AND a.deleted_at IS NULL
+  AND CASE
+    WHEN $assessments_group_by = 'Set' THEN a.assessment_set_id = $group_id
+    WHEN $assessments_group_by = 'Module' THEN a.assessment_module_id = $group_id
+  END
+  AND la.assessment_id IS NULL;
+
+-- BLOCK delete_bulk_lineitems
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.sql:131]: Nit: old table name; no need for `bulk`, just use the plural table name.

File: apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.ts [hunk 34]
```diff
@@ -68,6 +110,23 @@ router.get(
...
+        lti13_course_instance_id: req.params.unsafe_lti13_course_instance_id,
+      },
+      Lti13CombinedInstanceSchema,
+    );
+
+    const serverJobOptions = {
+      courseInstanceId: res.locals.course_instance.id,
+      userId: res.locals.user.user_id,
+      authnUserId: res.locals.authn_user.user_id,
+      type: 'lti13',
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorInstanceAdminLti13/instructorInstanceAdminLti13.ts:126]: Nit: consider using more precise job names so that we can e.g. show a history of all syncs, or something like that. That is, differentiate from syncs, create-and-link, etc.

**Review by nwalters512 — APPROVED** [review]
In the interest of letting you move on to grade passback, I won't block on any of these. I'm happy to do cleanup work on your behalf once this lands, just let me know.

File: apps/prairielearn/src/ee/lib/lti13.sql (hunk 29 above)
Comment (trombonekenny) [reply apps/prairielearn/src/ee/lib/lti13.sql]: Yeah, that's reasonable. Easy enough to fix, so I did.

**Comment by nwalters512** [comment]
Approval stands; merge when ready!
